File: 剑指OfferII038.每日温度/剑指OfferII038.每日温度.cpp
/*
 * @Descroption: 剑指 Offer II 038. 每日温度
 * @Author: EmoryHuang
 * @Date: 2021-08-04 19:47:35
 * @Method:
 * 维护一个单调栈
 * 遍历数组，若遇到比栈顶元素更大的元素，则更新出栈元素对应的等待天数
 * 并将更大的这个元素入栈
 */

class Solution {
   public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        stack<int> st;
        vector<int> ans(n);
        for (int i = 0; i < temperatures.size(); i++) {
            // 维护递减栈
            while (!st.empty() && temperatures[st.top()] < temperatures[i]) {
                // 更新等待天数
                ans[st.top()] = i - st.top();
                st.pop();
            }
            st.push(i);
        }
        return ans;
    }
};

File: Solution/剑指OfferII038.每日温度/剑指OfferII038.每日温度.java
/*
 * @Descroption: LeetCode 739. 每日温度
 * @Author: EmoryHuang
 * @Date: 2023-07-22 10:38:15
 * @Method:
 * 维护递减单调栈
 */

import java.util.*;

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        Deque<Integer> st = new LinkedList<>();
        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            while (!st.isEmpty() && temperatures[st.peek()] < temperatures[i]) {
                int idx = st.poll();
                ans[idx] = i - idx;
            }
            st.push(i);
        }
        return ans;
    }
}